Add setting for who can mention a user group.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First commit adds `group_setting_name` field to `GroupPermissionSetting`.
- Second commit is a prep commit refactoring tests.
- Next 5 commits are prep commits taken from #25884.
- Next commit is to add `can_mention_group` field to `UserGroup`.
- Next 2 commits are to add backend checks while sending and editing messages.
- Next 2 commits are to add API support for setting and changing `can_mention_group` setting.
- Last commit is to hide the groups not allowed to mention from typeahead.

Fixes part of #25927.

<!-- Issue link, or clear description.--> 

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented on zerver/lib/user_groups.py:
```diff
@@ -226,20 +226,46 @@ def get_recursive_subgroups_for_groups(user_group_ids: List[int]) -> List[int]:
+
+    for setting_name, permission_config in UserGroup.GROUP_PERMISSION_SETTINGS.items():
+        default_group = system_groups_name_dict[permission_config.default_group_name]
+        setattr(user_group, setting_name, default_group)
```
We can also pass a list of user groups also to this function and call `bulk_update` here instead of doing it in the caller of this function.

alexmv replied:
That feels like it would complicate `create_user_group_in_database` more.

sahil839 replied:
Yes true.

sahil839 replied:
Yes true.

timabbott replied:
Yeah this pattern seems mostly fine to me; the only database queries we do are in `get_role_based_system_groups_dict`.

I guess we're not calling this from `create_system_user_groups_for_realm`, but if we were to, we might want to refactor this in some way to avoid doing `get_role_based_system_groups_dict` in a loop.

sahil839 replied:
I did this refactoring to call `get_role_based_system_groups_dict` in this PR itself.

sahil839 commented on zerver/lib/import_realm.py:
```diff
@@ -969,6 +969,9 @@ def do_import_realm(import_dir: Path, subdomain: str, processes: int = 1) -> Rea
             re_map_foreign_keys(data, "zerver_usergroup", "realm", related_table="realm")
+            re_map_foreign_keys(
+                data, "zerver_usergroup", "can_mention_group", related_table="usergroup"
+            )
```
I am not sure whether we need to consider the case where the export data is old like the data contains user group but not `can_mention_group_id` field.

timabbott replied:
We do not -- we only support importing after exporting from the same version.

sahil839 commented on zerver/lib/message.py:
```diff
@@ -1633,6 +1635,19 @@ def wildcard_mention_allowed(sender: UserProfile, stream: Stream) -> bool:
+
+    for group in user_groups:
+        can_mention_group = group.can_mention_group
+        if not is_user_in_group(can_mention_group, sender, direct_member_only=False):
```
This does a DB query in a loop. But am not sure how often we would see a large number of user groups mentioned in a single message. I tried replacing these with a bulk query but that seemed somewhat complicated due to recursive groups being involved especially when we would like to show the ID of the group which user cannot mention. I can try again if we really want to call `is_user_in_group` in a loop.

alexmv replied:
Mmm, yeah -- this feels probably-fine to me, but I hear your worry.

Is `user_group_ids` ordered by the order of the mentions, so they at least get the message for the _first_ failing group, so we get some advantage over iterating over them?

sahil839 replied:
`mentions_user_group_ids` field of `MessageRenderingResult` object is a `set`, which I guess is unordered. Though this function receives a list as parameter which is obtained by applyig `list()` on the set and I did some print debugging and using `list()` probably always results in a increasing order of group IDs.

alexmv replied:
That feels like an implementation detail that we don't want to rely on.  Though "sorted by internal ID" isn't going to be differentiable by users from "sorted randomly," so I don't think there'd be much utility to sorting, here, other than generating a consistent error for the same set of groups.

If we decide we care about order, we could switch the `mentions_user_group_ids ` `set` to a `dict`, which would preserve insertion order.

sahil839 replied:
I don't think there is any real advantage of having `user_group_ids` in the same order as in the message. We cannot be sure at what position the user will mention the group they are not allowed to mention.

timabbott replied:
Hmm. I was going to say that this seems OK because you're limited in how much resources you can consume by the number of user groups on the server... but that could end up being a few hundred. I think we're probably OK for merging this, but cleaning this up should be a high priority follow-up.

I think the best implementation for this check would be to have a function similar in structure to `get_recursive_subgroups_for_groups` that can do the query "Is this user a member of all of these groups". One theory would be to define such a function in `zerver/lib/user_groups.py` that uses this inefficient implementation to let us integrate this PR, and then start a #backend thread about how to write that function efficiently.

sahil839 commented on zerver/actions/user_groups.py:
```diff
@@ -266,3 +268,20 @@ def check_delete_user_group(
+    # added here once that is merged.
+    setting_id_name = setting_name + "_id"
+    event_data_dict: Dict[str, Union[str, int]] = {setting_id_name: setting_value_group.id}
+    do_send_user_group_update_event(user_group, event_data_dict)
```
The event here is similar to the event send for name or description change of the user group. Do we want this to be something like we have for realm setting, where we have a `property` field set to setting name and `value` field set to the new setting value.

timabbott replied:
Let's discuss that in the "api design" stream.

sahil839 replied:
Started discussion - https://chat.zulip.org/#narrow/stream/378-api-design/topic/event.20for.20updating.20group.20setting

sahil839 commented:
Posted some inline comments above. And just to confirm, I am assuming that silent mention should not be restricted as per the new setting and should be allowed always.

alexmv commented on zerver/lib/import_realm.py:
```diff
@@ -961,49 +961,36 @@ def do_import_realm(import_dir: Path, subdomain: str, processes: int = 1) -> Rea
     realm_properties["deactivated"] = True
-    realm = Realm(**realm_properties)
 
-    if realm.notifications_stream_id is not None:
-        notifications_stream_id: Optional[int] = int(realm.notifications_stream_id)
-    else:
-        notifications_stream_id = None
-    realm.notifications_stream_id = None
-    if realm.signup_notifications_stream_id is not None:
-        signup_notifications_stream_id: Optional[int] = int(realm.signup_notifications_stream_id)
-    else:
-        signup_notifications_stream_id = None
-    realm.signup_notifications_stream_id = None
```
Can you clarify, in the commit mesasge, what made this block unnecessary?

sahil839 replied:
Thanks for the suggestion. This seems better. Updated the commit message.

alexmv commented on zerver/lib/message.py:
```diff
@@ -1633,6 +1635,19 @@ def wildcard_mention_allowed(sender: UserProfile, stream: Stream) -> bool:
+        can_mention_group = group.can_mention_group
+        if not is_user_in_group(can_mention_group, sender, direct_member_only=False):
+            raise JsonableError(
+                _("You cannot mention user group {user_group_id}").format(user_group_id=group.id)
```
In this case, do we want to tell them who _can_ mention them?  I guess that may be leaking information which is not-public.

Maybe `You are not allowed to` rather than `You cannot`?

sahil839 replied:
Updated the error message to use `You are not allowed to`. We are passing the `can_mention_group_id` field in the UserGroup objects sent to clients as we would require them to show the setting value in UI.

alexmv replied:
Ah, OK.  So we _could_ tell them "You must be a member of _Cool People_ or _Awesome People_ to mention this group."

sahil839 replied:
Do we want to mention group ID or group name here? We can also change the "You cannot mention .." part to use group name instead of ID to be consistent. 

Also, do we want to mention subgroups also in "You must be a meber of ..." part in the commit message? I think it should be sufficient to just mention the supergroup and it should be undestandable that it includes subgroups as well. Also a user group can have many subgroups, so the message would become too long in some cases.

timabbott replied:
Yeah, I think we only need to mention the supergroup - you're a member of the "Members role group" even if you're an administrator, just indirectly.

sahil839 replied:
Updated the error message to add "You must be a member of  ....." text.

alexmv reviewed (commented):
Is there a test you can add to the early commits about cache keys for deleted realms?  How are those commits related to the rest of the PR?

sahil839 commented:
Updated the PR with following changes and replied to the inline comments as well-
- Changed the error message as suggested above.
- Updated the commit message for the commit which moves part of `do_import_realm` inside a transaction.

sahil839 commented:
> Is there a test you can add to the early commits about cache keys for deleted realms? How are those commits related to the rest of the PR?

There has been discussion on this [here](https://chat.zulip.org/#narrow/stream/49-development-help/topic/Circular.20dependency.20of.20Django.20models/near/1584514). We need these changes because we add a new setting with `on_delete` set to [`RESTRICT`](https://docs.djangoproject.com/en/4.2/ref/models/fields/#django.db.models.RESTRICT) which apparently results in `Realm` objects being deleted first and then `AlertWord`, `RealmEmoji`, etc. objects are deleted. Due to this, when the `post_delete` signals are called for `AlertWord` and similar objects, the `realm` passed there was already deleted, so the commits changes them to accept `realm_id` as parameter, since we anyways need only the ID.

I am not sure what and how to test here. The cache keys remain the same we just pass `realm_id` instead of `realm` to the function which returns cache key. Do we want to test that the cache is actually flushed or am I missing something?

alexmv commented:
> I am not sure what and how to test here. The cache keys remain the same we just pass realm_id instead of realm to the function which returns cache key. Do we want to test that the cache is actually flushed or am I missing something?

No, after reading the topic, I think this is fine -- the tests would start failing due to the `RESTRICT` change, and these changes are effectively necessary to handle that.

I guess something more in the commit messages might clarify -- maybe something along the lines of:

> Subsequent commits will add `on_delete=models.RESTRICT` relationships, which can cause objects to be deleted after their Realm has been deleted from the database.  In order to handle this, we switch caching to be by `id`, which will continue to work even if the Realm object itself has already been deleted.

sahil839 commented:
Thanks for the suggestions for commit message for cache-related commits. Updated the commit messages to be explain the reason of change more clearly.

sahil839 commented:
Last push was to update the import realm code to handle all the group settings by using `GROUP_PERMISSION_SETTINGS` dict.

timabbott commented:
The prep commits look great; created https://github.com/zulip/zulip/pull/26149 to run CI on those and merge them. @Ujjawal3 FYI since many of them are from one of your PRs.

timabbott commented on zerver/migrations/0455_set_default_for_can_mention_group.py:
```diff
@@ -0,0 +1,38 @@
+
+    groups_to_update = []
+    for realm in Realm.objects.all():
+        # The default value of `can_mention_group` is everyone system_group.
```
I guess to be concrete, this implements the ability to mention role system groups? I think there was an issue for that...

timabbott commented on zerver/lib/message.py:
```diff
@@ -1633,6 +1635,21 @@ def wildcard_mention_allowed(sender: UserProfile, stream: Stream) -> bool:
+        can_mention_group = group.can_mention_group
+        if not is_user_in_group(can_mention_group, sender, direct_member_only=False):
+            raise JsonableError(
+                _("You are not allowed to mention user group {user_group_id}").format(
```
I think the name is probably going to be a lot more helpful to users, since this seems like an error you might encounter in the web app.

sahil839 replied:
Updated this to mention group name instead of ID.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -16320,6 +16321,12 @@ paths:
+                              description: |
+                                ID of the user group whose members are allowed to mention the group.
+
+                                **Changes**: New in Zulip 8.0 (feature level 190).
```
We should spell out "Previously, groups could be mentioned if any only if they were not system groups." in the changes line.

sahil839 replied:
Added this line.

timabbott commented on zerver/views/user_groups.py:
```diff
@@ -98,8 +99,9 @@ def edit_user_group(
 ) -> HttpResponse:
-    if name is None and description is None:
+    if name is None and description is None and can_mention_group_id is None:
         raise JsonableError(_("No new data supplied"))
```
I think we've been moving away from having checks like this in API endpoints; we should discuss removing this check as a potential API change to this endpoint in "api design".

sahil839 replied:
Started discussion - https://chat.zulip.org/#narrow/stream/378-api-design/topic/PATCH.20.2Fuser_groups.2F.7Buser_group_id.7D

timabbott commented on web/src/user_groups.ts:
```diff
@@ -88,6 +89,21 @@ export function get_realm_user_groups(): UserGroup[] {
+        return [];
+    }
+
+    const user_groups = get_realm_user_groups();
```
I guess I note that we're still hiding system groups here.

timabbott commented on web/tests/composebox_typeahead.test.js:
```diff
@@ -1533,18 +1533,38 @@ test("content_highlighter", ({override_rewire}) => {
+
+    // call_center group is not shown in typeahead even when user is neither
+    // a direct member of can_mention_group nor a member of any of its
+    // recursive subgroups.
```
I'm confused by the comments here; is "even when" a typo?

timabbott commented:
Overall, this looks pretty good. I think there's something a bit weird with making the existing system groups mentionable -- we're not doing it in the frontend, but are doing it in the backend ... and more awkwardly, their names starting with `@` are certainly not designed for mentions. If that ends up being a tangle, it would be an option to not try to address #23269 in this PR, and have the mentionable flag we initialize in the migration and in realm creation be `@role:nobody` rather than `@role:everyone` for system groups.

sahil839 commented on zerver/lib/types.py:
```diff
@@ -281,3 +281,4 @@ class GroupPermissionSetting:
     allow_owners_group: bool
     allow_nobody_group: bool
     default_group_name: str
+    default_for_system_groups: Optional[str] = None
```
Have added this since I guess we might have settings in future that might need different default for system groups. We would have the same default for all groups atleast for this setting once we solve #26148, but then also we might have different default for single user groups.

Might want to rename this or add a new field then in future depending upon what further group-level settings we add. 
Or we can probably skip this for now as we only have one setting currently, not sure.

Have kept this optional since this will not be used for stream and realm setings. Though we might want to have different default for public and private streams but that can be handled in future.

timabbott replied:
Yeah, this seems pretty reasonable for now.

timabbott commented:
This looks great, merged, thanks for getting this put together so quickly @sahil839!
